### PR TITLE
CFEP-06: staged-recipes code review lifecycle (Automating staged-recipes PRs Part 1)

### PR DESCRIPTION
**Pull request by ericdill**

### CODE REVIEW

**Comment by ericdill** [comment]
Hi @conda-forge/core and @conda-forge/staged-recipes,

On today's dev meeting (2018-07-24) we talked about automatically closing pull-requests (PRs) on
staged-recipes to keep the number of open PRs manageable for the @conda-forge/core and
@conda-forge/staged-recipes teams. Right now we have almost 400 open PRs. We are somewhere
between "many" and "most" of those never being merged. I think it is time to take action and
clean up our stale PRs on staged-recipes.

**Comment by mwcraig** [comment]
This looks reasonable (if a little complicated). I agree the old PRs need to get closed out.

As another example, what was implemented in astropy was a bot that, after N months of inactivity, posts a comment saying that the PR will be closed in Y month(s) automatically unless a particular label is added to the PR.

For astropy we went with N=5, Y=1, IIRC. 

I gather all of the label changing here is automated; if not, I think it is too complicated to stick with in practice.

**Comment by ericdill** [comment]
Thanks for the comments @mwcraig. Do you have any suggestions for simplifying the workflow here?

**Comment by jjhelmus** [comment]
> As another example, what was implemented in astropy was a bot that, after N months of inactivity, posts a comment saying that the PR will be closed in Y month(s) automatically unless a particular label is added to the PR.

I do not believe pull requests authors have the ability to add, modify or remove a label from a PR.  This can results in an unfortunate situation where an author updates an PR, no one responds and the bot closes the PR.  This is a situation I would like to avoid.  A bot should never close a PR where the last action was by the author.

**Comment by djsutherland** [comment]
Agreed, labels are an issue. What if the author needs to reply with something like `@conda-forge-admin This is still relevant.` to avoid auto-closing?

**Comment by isuruf** [comment]
Or any comment would remove the stale label. (this is what https://probot.github.io/apps/stale/ does)

**Comment by mwcraig** [comment]
@ericdill -- I took another look after getting some sleep; I don't see any easy simplifications and gather the whole thing would be automated anyway.

I do like the "awaiting review" label; I only occasionally have time to review recipes, but when I do at least 15 min seems to go into finding PRs that are really only awaiting review and I think this will help.

File: cfep-06.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+* Can be closed
+* Closed
+* Merged
+
+We have 10 proposed transitions. The syntax here is `+` or `-` followed by a label. This indicates
+that a specific label is added or removed.
+
+0. **Author** submits pull-request to staged-recipes
+    * +"Awaiting Review"
+1. A **Reviewer** reviews the recipe and requests changes.
```
Comment (mwcraig) [inline cfep-06.md:93]: This seems to imply (and probably *should* imply) that reviews should either "Request changes" (in the github PR review sense) or "Approve".  If any feedback should trigger "Awaiting Author" then maybe change the wording here to "reviews the recipe".

As a corollary, should reviews be instructed to *not* use the "Comment" option when reviewing?

**Comment by mwcraig** [comment]
> I do not believe pull requests authors have the ability to add, modify or remove a label from a PR. This can results in an unfortunate situation where an author updates an PR, no one responds and the bot closes the PR. This is a situation I would like to avoid. A bot should never close a PR where the last action was by the author.

Agreed, but if I understand the proposal correctly the only way for a PR to get to 'Stale' or 'Can be closed' is by way of 'Awaiting Author', so only if the last action was by a reviewer.

I worry a bit that 'Can be closed" may never actually be closed unless it is automated. 

Even if we don't auto-close, I think having the bot *comment* on the PR (in addition to changing the label) would be helpful because it will trigger a notification to the PR author. 

I like the "voice" of the astropy bot regarding these things ([example](https://github.com/astropy/astropy/pull/6065#issuecomment-406299150)) even though it is wordy.

Suggestion for a 'stale' comment message:

----

Hi contributor! I'm the conda-forge bot 🤖and want to encourage you to respond to the changes the staged-recipes humans requested. The humans would really like to merge this pull request soon! If you need help, please copy and paste this into a comment on this pull request: `Hey @conda-forge/staged-recipes, please help me with this pull request` 

Please update the pull request within the next month; if no changes have been made by then I'll give you another reminder.  

-----

Suggestion for a 'Can be closed' bot comment:

----

Hi contributor! I'm the conda-forge bot 🤖and want to encourage you to respond to the changes the staged-recipes humans requested. The humans would *really* like to merge this pull request, but need you to make some changes. If you need help, please copy and paste this into a comment on this pull request: `Hey @conda-forge/staged-recipes, please help me with this pull request` 

The pull request is likely to be closed within the next month if there are no updates....that would make the humans sad, so please help them out by wrapping this awesome pull request up!

-----

**Comment by beckermr** [comment]
This CFEP came up again today. I support it but don't have time to contribute. I'd say any PR older than a year can be automatically closed with a nice message.

**Comment by saraedum** [comment]
I was just added to the review team for staged-recipes and I find it too hard to figure out which PRs currently need review:

Basically, the best way seems to be to look for pings to staged-recipes (or similar), but these can only be sent by conda-forge members. Also, there seems to be no easy way to see whether somebody else has already reacted to that ping without browsing through the PR quickly.

So anything that serves as a "awaiting review"/"awaiting author" label would be a great improvement. So, this proposal sounds good to me (though I don't get to vote here.)

**Comment by saraedum** [comment]
A technically trivial solution that could be implemented right away would be to add `[awaiting review]` or `[awaiting author]` to the title of a PR to indicate its state. While authors cannot change labels, they can change the title, and reviewers can search `review in:title` to find the ones needing review quickly.

**Comment by jakirkham** [comment]
Is this something you would be interested in championing @saraedum? 🙂

**Comment by saraedum** [comment]
@jakirkham sure. There was a bit of discussion on gitter about this at some point. It was not clear what the process would be. An actual CFEP or just a PR to staged-recipes to modify the PR template?

**Comment by jakirkham** [comment]
This CFEP was trying to develop a workflow around `staged-recipes` PR review. This would be useful in just articulating what the intended workflow is. It sounds like you have thoughts on/interest in it. So the CFEP seems like the natural place to start (given there is already work on it). That said, I don't want to dictate a process to you as you likely have a good idea on how we should start clarifying the review process.

**Comment by saraedum** [comment]
Sorry, @jakirkham I maybe misunderstood. Did you mean champion "this" CFEP or champion what I had written before:
> A technically trivial solution that could be implemented right away would be to add [awaiting review] or [awaiting author] to the title of a PR to indicate its state. While authors cannot change labels, they can change the title, and reviewers can search review in:title to find the ones needing review quickly.

If the latter, then I am not sure what the process would be. Create a new CFEP or can this be handled by just some informal discussion on a PR on staged-recipes?

File: cfep-06.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
+
+<table>
+<tr><td> cfep-06 </td><td> staged-recipes code review lifecycle </td>
+<tr><td> Status </td><td> **Draft** | Proposed | Accepted | Rejected | Deferred | Implemented </td></tr>
```
Comment (beckermr) [inline cfep-06.md]:
```suggestion
<tr><td> Status </td><td> Deferred </td></tr>
```

**Comment by beckermr** [comment]
merging as deferred

**Pull request merged**